### PR TITLE
Bugfix - Client Steering EasyMesh Messages

### PR DESCRIPTION
**Pull request by vitalybu**
This PR fixes the client steering logic after switching to EasyMesh messages.
Current implementation sends incorrect values and missing local agent BSSID filtering.
This caused the target agent to add the steering station into it's DENY list which completely broke the mechanism.

### CODE REVIEW

**Review by tomereli — APPROVED** [review]

**Review by tomereli — CHANGES_REQUESTED** [review]
please explain in the PR description though how this was tested and on which target.
Also, it looks like the implantation for nl80211 BWL is not taking the bssid into account. Perhaps we can filter the bssid in the ap manager instead of the BWL itself? As a matter of fact, even the controller has the vap list so you can save the messages to the other agents.

**Review by tomereli — COMMENTED** [review]
Not approving yet so Arnout has also a chance to review.

**Comment by vitalybu** [comment]
> please explain in the PR description though how this was tested and on which target.
> Also, it looks like the implantation for nl80211 BWL is not taking the bssid into account. Perhaps we can filter the bssid in the ap manager instead of the BWL itself? As a matter of fact, even the controller has the vap list so you can save the messages to the other agents.

This fix was tested on CGR running RDKB.
I thought about the right place to implement the filtering and decided to go with BWL for the following reasons:
- We can't assume anything about 3rd party controllers, so the filtering must be done in the agent.
- Currently the Agent and the AP Manager are somewhat agnostic to the underlying networking structure. They don't even store a list of VAPs. We can change this of course, and perhaps as more and more radio/vap specific messages will be implemented it'll make more sense, but at this point I came to the decision that BWL filtering is good enough.
- BWL is platform specific, and some platform may need to handle this differently.

As for the NL80211 BWL implementation, the BSSID is indeed ignored there, but it doesn't break the implementation. The fix was made in the controller, so any NL80211 platform will keep Allowing/Disallowing stations for the entire radio (we don't currently support any NL80211 platforms with VAPs anyway).

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:140]: If you remove ipv4, then the same commit should also remove its uses from son_slave_thread and ap_manager_thread, otherwise the build breaks.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_1905_vs.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_1905_vs.yaml:40]: At this point, it *is* actually still used in slave_thread::handle_client_association_request, so that has to be removed first.

File: common/beerocks/bwl/include/bwl/ap_wlan_hal.h
Comment (arnout) [inline common/beerocks/bwl/include/bwl/ap_wlan_hal.h:111]: I don't want any new code that uses std::string instead of sMacAddr. Ideally there would be a separate commit that changes the mac to sMacAddr as well, but for the time being I'm OK with just adding `const sMacAddr &bssid`.

The changes to the caller have to be done in the same commit. To make the transition smoother, you could give the bssid a default value of ZERO_MAC here and remove that default later.

**Comment by tomereli** [comment]
> > please explain in the PR description though how this was tested and on which target.
> > Also, it looks like the implantation for nl80211 BWL is not taking the bssid into account. Perhaps we can filter the bssid in the ap manager instead of the BWL itself? As a matter of fact, even the controller has the vap list so you can save the messages to the other agents.
> 
> This fix was tested on CGR running RDKB.
> I thought about the right place to implement the filtering and decided to go with BWL for the following reasons:
> 
> * We can't assume anything about 3rd party controllers, so the filtering must be done in the agent.
> * Currently the Agent and the AP Manager are somewhat agnostic to the underlying networking structure. They don't even store a list of VAPs. We can change this of course, and perhaps as more and more radio/vap specific messages will be implemented it'll make more sense, but at this point I came to the decision that BWL filtering is good enough.
> * BWL is platform specific, and some platform may need to handle this differently.
> 
> As for the NL80211 BWL implementation, the BSSID is indeed ignored there, but it doesn't break the implementation. The fix was made in the controller, so any NL80211 platform will keep Allowing/Disallowing stations for the entire radio (we don't currently support any NL80211 platforms with VAPs anyway).

Good points, thank you for the explanation. I'll approve, but let's give @arnout also a chance to review before merging?

**Review by arnout — APPROVED** [review]
The commit splitting/ordering is not OK, but that's nitpicking.

I really don't want new uses of string for MAC address. However, since the functions in bwl are already using strings, it's also something that can be fixed later.

I also think the slave should have a copy of the VAP list. But even if it does, the BSSID has to be passed to the AP manager so this PR still applies.

Note that we will also need to handle the validity period of the blocking request at some point. However, this is not part of the test plan, so not included in #405 

Another thing that I'm wondering about is: wouldn't it be better if we sent the wfa_map tlvClientAssociationControlRequest as-is to the AP Manager? Why do we need to do this conversion in the slave?

**Review by tomereli — APPROVED** [review]
Arnout raised valid comments, so consider fixing those before merging.

**Comment by vitalybu** [comment]
> The commit splitting/ordering is not OK, but that's nitpicking.
> 
> I really don't want new uses of string for MAC address. However, since the functions in bwl are already using strings, it's also something that can be fixed later.
> 
> I also think the slave should have a copy of the VAP list. But even if it does, the BSSID has to be passed to the AP manager so this PR still applies.
> 
> Note that we will also need to handle the validity period of the blocking request at some point. However, this is not part of the test plan, so not included in #405
> 
> Another thing that I'm wondering about is: wouldn't it be better if we sent the wfa_map tlvClientAssociationControlRequest as-is to the AP Manager? Why do we need to do this conversion in the slave?

I agree that the commit splitting and ordering is not optimal, but in this case I preferred to split in this way for easier review (assuming that the stand-alone commits won't be cherry-picked anywhere).

Thought about passing the `tlvClientAssociationControlRequest` message as-is to the AP Manager, but currently the AP Manager handles vendor-specific messages [only](https://github.com/prplfoundation/prplMesh/blob/bugfix/client-steering-easymesh-messages/agent/src/beerocks/slave/ap_manager_thread.cpp#L425). Should probably be changed later.

I'll merge this PR now to unblock work on UGW8.x and RDKB, but let's discuss the Agent<->AP Manager messaging mechanism in the next architectural meeting.

**Pull request merged**
